Deploy site (#3)
raresdolga/personal-blog: _posts/2024-04-14-latte.md

---
title: "Latte: Latent Attention for Linear Time Transformers"
date: 2024-04-14
layout: page
---

<figure style="text-align: center;">
  <img
     src="{{'/assets/img/bid_latte.png' | relative_url }}"
     alt="bid_late"
     class="img-responsive"
     style="float: center;margin-left: 35%;"
    >
  <!--figcaption>MDN Logo</figcaption-->

> </figure>

Reviewer: More of a question than a comment, but should the lines in this figure have arrows?

Author: I do not think so. It just aims to show which elements interact, there is no ordering.  Happy to make some changes if we can make this thing clear, but might confuse people more with arrows

We introduce Latte, a new linear time and memory attention mechanism which, unlike other sparse methods, takes all the tokens into account. Our method can be used as a replacement for the standard attention in both bidirectional and causal settings with similar empirical performance. An additional benefit of causal Latte is efficient inference, which only requires constant time to predict the next token compared to the linear scaling requirement of the standard attention.

> We are first going to rewrite the classic attention mechanism in the non-vectorises form, which will help us describe the idea behind Latte more intuitively.

Reviewer: ```suggestion
We are first going to rewrite the classic attention mechanism in the non-vectorised form, which will help us describe the idea behind Latte more intuitively.
```

<h2>Quick Recap</h2>

One of the most common ways of writing an attention layer is using the matrix form:
 \[

>  Softmax(\frac{1}{\sqrt{d}}QK^T)V

Reviewer 2: Nit: I think you can do
```
\text{Softmax} \big(( \frac{1}{\sqrt{d}}QK^T \bigg) V
```
Here to make `Softmax` appear as a function eg. `exp` and to match the softmax brackets to the `\frac`

Reviewer: +1 - was also about to comment on the bracket sizes.

I'm not sure about GitHub LaTeX, but I'm not 100% sure you need the `bigg`? i.e. in most LaTeX I think you can just do `\( ... \)`.

Reviewer: Also would use `\mathsf{T}` ($\mathsf{T}$) for the transpose if it's available.

Reviewer: Okay just tested it, I think you do need the `bigg`. `\text{Softmax} \bigg( \frac{1}{\sqrt{d}}QK^\mathsf{T} \bigg) V`

$\text{Softmax} \bigg( \frac{1}{\sqrt{d}}QK^\mathsf{T} \bigg) V$

Author: Yeah. I could have used \intercal as well, but it is ugly. Thanks both!

 \]

> Nonetheless, the formula can be written more intuitively without any vectorisation. Considering a sequence of $T$ tokens from which we obtain the query $q_i$, key $k_i$ and value $v_i$ we can explain attention using the figure below:

Reviewer: Maybe just me being pedantic (in which case feel free to ignore) but I'm not a fan of indexing a variable with $i$ and then having it go up to $T$. I'd use $t$ as the index.

Author: You are right

> <figure style="text-align: center;">

Reviewer 2: I really like this figure!

  <img
     src="{{'/assets/img/att.png' | relative_url }}"
     alt="Attentention Non-Vect"
     class="img-responsive"
     style="float: center;"
    >
    <br>
  <figcaption>Figure1: Non vectorised attention mechanism</figcaption>
</figure>

In the above, we defined:
\[A_{ij} = \frac{\exp(Q_{i,:}K_{:,j}^{T})}{\sum_{j=1}^T \exp(Q_{i,:}K_{:,j}^{T}) }\]
and
\[
\tilde{x_i} = \sum_{j=1}^T A_{ij}V_j

> \]

Reviewer: I think it would be good to introduce what $A$ and $\tilde{x}$ are before (or immediately after) using them in equations.

For the causal case we would simply sum up to $i$ instead of $T$ such that tokens in the future are not considered. 

<h2> Latte Layer </h2>

> The problem with attention is that computing all $A_{ij}$ is quadratic in sequence length, hence we need to find a more efficient way to summarise the interactions between tokens. One simple approach is to introduce a few learnable tokens and to calculate attention only between the introduced tokens and the original sequence. This approach was successfully explored by many sparse methods like [BigBird](https://arxiv.org/abs/2007.14062), however in the new representation $\tilde{x_i}$ only the learned tokens will contribute.

Reviewer: I don't think it's particularly clear what is meant by "only the learned tokens will contribute" - maybe you could expand on this a bit.

<figure style="text-align: center;">
  <img
     src="{{'/assets/img/comp_att.png' | relative_url }}"
     alt="Comparison Attentions"
     class="img-responsive"
     style="float: center;"
    >
  <figcaption>Figure2: Comparison between Bidirectional Standard attention, sparse attention methods and Bidirectional Latte.</figcaption>
</figure>

> More exactly, we replace each $A_{ij}$ with an approximation based on $L$ learnale latent variables:

Reviewer 2: Nit: `learnale` -> `learnable`

> More exactly, we replace each $A_{ij}$ with an approximation based on $L$ learnale latent variables:

Reviewer: ```suggestion
Specifically, we replace each $A_{ij}$ with an approximation based on $L$ learnale latent variables:
```

\[
A_{ij} = \sum_{l=1}^L Q_{il}K_{lj}
\]  
The new vector representation of each token is given by:
\[
\tilde{x_i} = \sum_{l=1}^L Q_{il} \sum_{j=1}^T K_{lj}V_j
\]
Note that $Q$, $K$ have different sizes than the queries and keys in the standard attention. Figure 3 describes in detail how we obtain these matrices.

<figure style="text-align: center;">
  <img
     src="{{'/assets/img/fig2.png' | relative_url }}"
     alt="Latte Implementation"
     class="img-responsive"
     style="float: center;"
    >
    <br>
  <figcaption>Figure 3: Each row in X corresponds to a token in the input sentence. We multiply the input by the corresponding weights and obtain Q,K,V. Note that we exponentiate and normalize both Q and K on rows.</figcaption>
</figure>

> The approach is motivated by a probabilistic perspective where $A_{i,:}$ is interpreted as a probability distribution. While in this blog we describe the model at an intuitive level, a more rigorous mathematical explanation can be found in <a href="https://arxiv.org/abs/2402.17512" target="_blank">our paper</a>

Reviewer: ```suggestion
The approach is motivated by a probabilistic perspective where $A_{i,:}$ is interpreted as a probability distribution. While in this blog we describe the model at an intuitive level, a more rigorous mathematical explanation can be found in <a href="https://arxiv.org/abs/2402.17512" target="_blank">our paper</a>.
```
